app/javascript/utilities/protectedFetch.js

```diff
@@ -0,0 +1,40 @@
+import {authenticatedFetch} from '@shopify/app-bridge-utils';
+import {AuthCode} from '@shopify/app-bridge/actions';
+
+function getAuthCode(appBridge) {
+  return new Promise((resolve, reject) => {
+    const unsubscribe = appBridge.subscribe(AuthCode.ActionType.RESPOND, (payload) => {
+      if (payload) {
+        resolve(payload);
+      } else {
+        reject("Failed to load");
+      }
+      unsubscribe();
+    });
+
+    appBridge.dispatch(AuthCode.request());
+  });
+}
+
+function invalidUserAccessTokenError(responseObject) {
+  return responseObject.headers.get("X-Shopify-Request-Auth-Code");
+}
+
+export default function protectedFetch(appBridge) {
+  const jwtFetch = authenticatedFetch(appBridge);
+
+  return async (uri, options) => {
+    const response = await jwtFetch(uri, options);
+    if (invalidUserAccessTokenError(response)) {
+      const { code, hmac, shop, timestamp } = await getAuthCode(appBridge);
+      const callback_uri = `auth/shopify/callback?code=${code}&hmac=${hmac}&shop=${shop}&timestamp=${timestamp}`
+
+      const callback_response = await jwtFetch(callback_uri);
+      if (callback_response.status == 200) {
```

Comment:
One thing to be concerned about is if fetching the token is handled synchronously or not. If it is async on the backend, then you might have to introduce a slight delay here.

Reply:
From what we can tell, it is synchronous as there are no redirects and it sends back a single response with the status.